Init

hanny24 opened the pull request.

Commit:
initial

Commit:
Merge branch 'master' into init

# Conflicts:
#	README.md

Commit:
example outputs

Commit:
typo

sideeffffect commented on README.md:
```diff
@@ -1 +1,23 @@
+   * line number
+   * loglevel as both string and number
+ * is extensible: very little is assumed
+ * uses minimal dependencies
```
to the extent we don't reinvent the wheel and reuse existing things

I wouldn't want us to re-implement a poorer, less configurable and maintained clone of `slf4j` / `logback`

hanny24 replied:
Yes, implementation will be based on slf4j & logback.

sideeffffect commented on core/src/main/scala/slog/LogBuilder.scala:
```diff
@@ -0,0 +1,7 @@
+trait LogBuilder[F[_]] {
+  def msg(msg: String): F[Unit]
+  def msg(ex: Throwable)(msg: String): F[Unit]
+  def context[T: StructureEncoder](key: String, value: => T): LogBuilder[F]
```
I don' like the separation between creating a context and logging things. `context` and `use` do kinda the same thing.

hanny24 replied:
The point is to allow multiple context to be used.

sideeffffect replied:
I don't understand... you can call `context` multiple times, right?

hanny24 replied:
Ahh, there was a misunderstanding. I kind of didn't like to use `loggingContext.context`.

sideeffffect replied:
I would like those methods which have signature like `(key: String, value: T)` have the same name. Could it be `with`?
I don't personally care much about which specific name is chosen...

hanny24 replied:
`with` is a keyword, try something else:)

sideeffffect commented on example/src/main/scala/example/Example.scala:
```diff
@@ -0,0 +1,66 @@
+          logger.debug.msg(new Exception)("test") >>
+          // {"message": "test2", "file": "Example.scala", "line": 35, "correlation_id": "<VALUE>", "request_id": "<VALUE>"}
+          logger.info.msg("test2")
+        }
```
here I try to illustrate how I would like the API to change
```scala
    logger.info("Hello world!!") >>
      // Provide extract context to the log message
      // {"message": "Hello", "file": "Example.scala", "line": 22, "correlation_id": "<VALUE>"}
      logger.context("correlation_id", "<VALUE>").info("Hello") >>
      // Provide extra context and an exception
      // {"message": "Hello", "file": "Example.scala", "line": 25, "correlation_id": "<VALUE>", "stack_trace": "Exception ..."}
      logger
        .context("correlation_id", "<VALUE>")
        .info(new Exception)("Hello") >>
      logger
        .context("correlation_id", "<VALUE>")
        .context("request_id", "<VALUE>")
        .scoped {
          // {"message": "test", "file": "Example.scala", "line": 33, "correlation_id": "<VALUE>", "request_id": "<VALUE>", "stack_trace": "Exception ..."}
          logger.debug(new Exception)("test") >>
          // {"message": "test2", "file": "Example.scala", "line": 35, "correlation_id": "<VALUE>", "request_id": "<VALUE>"}
          logger.info("test2") 
```

hanny24 replied:
I'm OK with switching `context` and log level. What I'm not OK is usage of `scoped` directly on `logger`. This somehow suggest that it is scoped only for *that particular logger*, which is not the case as it is scoped to the all loggers.

jakubjanecek replied:
How about calling it just `scope` instead of `scoped`? Meaning "I am creating a scope...".

sideeffffect replied:
> scoped only for _that particular logger_

I mean, it could be the case. It depends on the other Loggers, if they share the same underlying "TaskLocal", and that really depends on how the programmer sets up the application

hanny24 replied:
Which is exactly why I want it to be separated. Because implementation should not really change semantics of the context.

sideeffffect replied:
I'm not sure agree, I think this should be driven by how programmer wires it up -- which usually would be just 1 TaskLocal

jveselka replied:
@hanny24 Good point about how calling scope (on logger instance) affects all loggers. Natural solution seems to have `object LoggingContext`. I see that it is `trait` passed as parameter, how do you propose multiple LoggingContexts would interect?

sideeffffect commented on generic/src/main/scala/slog/generic/auto.scala:
```diff
@@ -0,0 +1,17 @@
+package slog.generic
```
I have serious doubts about somewhat reinventing the wheel by having custom encoders. I'm not sure the benefit outweigh the costs compared to using directly Circe.

But we would have to discuss this in person with a wider group of stakeholders

jveselka replied:
I am strongly in favor of custom encoders. I think we should be able to provide implicit conversion from `io.circe.Encoder[A]` to `slog.StructureEncoder[A]` satisfying circe users while not being locked into it.

tomasherman replied:
i agree with custom format but against implicit conversion from circe - just because you can use something in API doesn't mean you should be able to log it in a structured way, imho - you should make an explicit decision about that

jveselka replied:
The decision to import this implicit conversion would still be up to user. I believe that is explicit enough.

sideeffffect commented on core/src/main/scala/slog/LogBuilder.scala:
```diff
@@ -0,0 +1,7 @@
+trait LogBuilder[F[_]] {
+  def msg(msg: String): F[Unit]
+  def msg(ex: Throwable)(msg: String): F[Unit]
+  def context[T: StructureEncoder](key: String, value: => T): LogBuilder[F]
```
this could come handy too
```scala
def context[T: StructureEncoder](m: Map[String, T]): LogBuilder[F]
```
or even better something kinda like
```scala
def context[T: StructureEncoder](m: Map[String, Lazy[T]]): LogBuilder[F]
```
where I don't yet know where to take `Lazy` from :)

hanny24 replied:
I don't think it's as useful as you think. Reason for that is you usually want to use different type for each key in the map.

sideeffffect replied:
I think it would be used rarely, but when that time comes... but this is a minor discussion point anyway

tomasherman replied:
if that time comes, you can move the entire map into `T`, no?

sideeffffect replied:
when I think of it, the method accepting `Map[String, T]` would rise in significance if we decided not to use the fluent/builder pattern and used flat methods instead

sideeffffect replied:
@tomasherman but then it would be nested one level lower, which might not be what you want. but it's a detail now

jveselka replied:
We could (once again :-) ) also provide implicit conversions from `StructureEncoder[T]` to `StructureEncoder[Map[String, T]]`. Regarding nesting, I think most would not care much.

jakubjanecek commented:
I looked at it and I like it! :slightly_smiling_face: Few things:
* Do we really need the concept of `LoggerFactory`? Couldn't we just include the `implicit classTag` everywhere? Or make it similar to Metrics with something like `Logger.named` method which allows you to name your `Logger`? I know having the factory is common but to me it seems like unnecessary concept and what I don't like about it is that you pass the factory and you create a "dependency" (the logger) in the class itself instead of passing the logger directly to constructor. Just an idea, what do you think?
* The macro for logging should do the log level check `if (isDebug)` but I guess that's just not finished yet.
* Do I understand it correctly that using `LoggingContext` and `logger.context` is essentially the same thing but `LoggingContext` allows the scoping? Again, do we need the `LoggingContext` and do we need to pass it as extra parameter to the class? It seems to me that this will be very cumbersome because you'll have like 2 or 3 constructor parameters yourself and then you'll have two for logging, one for monitoring and who knows what... I'd try to simplify this.
* Is there specific reason for the syntax `log.info.msg`? It's a bit unusual to me, I'm more used to `log.info(msg)` or something like that.

sideeffffect commented on example/src/main/scala/example/Example.scala:
```diff
@@ -0,0 +1,66 @@
+    logger.info.msg("Hello world!!") >>
+      // Provide extract context to the log message
+      // {"message": "Hello", "file": "Example.scala", "line": 22, "correlation_id": "<VALUE>"}
+      logger.info.context("correlation_id", "<VALUE>").msg("Hello") >>
```
calling `context` should go _before_ calling `info`
(the current order suggests there are different levels of context (`info`, `warn`,...), but that's not the case)

hanny24 replied:
> I looked at it and I like it! 🙂 Few things:
> 
>     * Do we really need the concept of `LoggerFactory`? Couldn't we just include the `implicit classTag` everywhere? Or make it similar to Metrics with something like `Logger.named` method which allows you to name your `Logger`? I know having the factory is common but to me it seems like unnecessary concept and what I don't like about it is that you pass the factory and you create a "dependency" (the logger) in the class itself instead of passing the logger directly to constructor. Just an idea, what do you think?

Well we don't need it, yet I didn't want to get too far from how logging is done right now. I'm actually not against it as long as there is big enough consensus on this.

>     * The macro for logging should do the log level check `if (isDebug)` but I guess that's just not finished yet.

It turns out it's not that easy if we want to be really pure. Current API tries to defer as much as possible, but there is still some allocation. And I'm not sure if macros will help in this case. (probably not)

>     * Do I understand it correctly that using `LoggingContext` and `logger.context` is essentially the same thing but `LoggingContext` allows the scoping? Again, do we need the `LoggingContext` and do we need to pass it as extra parameter to the class? It seems to me that this will be very cumbersome because you'll have like 2 or 3 constructor parameters yourself and then you'll have two for logging, one for monitoring and who knows what... I'd try to simplify this.

They look similar and even work similarly, but they are two different things. `LoggingContext` is used for context that is not really tied to a single log message. Typical example would be correlation id. On the other hand, `Logger.context` is used for log message specific context. That would typically be some case class.

It might seem that `Logger.context` is not really needed (which in fact is not), but I included it to simplify it for the user. And most importantly, it will allow us to provide more efficient implementation of this use case.

>     * Is there specific reason for the syntax `log.info.msg`? It's a bit unusual to me, I'm more used to `log.info(msg)` or something like that.

Not really, except that I found it more consistent with the rest.

jveselka commented on core/src/main/scala/slog/LoggingContext.scala:
```diff
@@ -0,0 +1,16 @@
+package slog
+
+trait LoggingContext[F[_]] {
+  def use[T: StructureEncoder, R](key: String,
```
What is `R` type parameter for?

hanny24 replied:
It was a leftover from a previous version. Removed. Thanks!

hanny24 commented:
> I looked at it and I like it! 🙂 Few things:
> 
>     * Do we really need the concept of `LoggerFactory`? Couldn't we just include the `implicit classTag` everywhere? Or make it similar to Metrics with something like `Logger.named` method which allows you to name your `Logger`? I know having the factory is common but to me it seems like unnecessary concept and what I don't like about it is that you pass the factory and you create a "dependency" (the logger) in the class itself instead of passing the logger directly to constructor. Just an idea, what do you think?

Well we don't need it, yet I didn't want to get too far from how logging is done right now. I'm actually not against it as long as there is big enough consensus on this.

>     * The macro for logging should do the log level check `if (isDebug)` but I guess that's just not finished yet.

It turns out it's not that easy if we want to be really pure. Current API tries to defer as much as possible, but there is still some allocation. And I'm not sure if macros will help in this case. (probably not)

>     * Do I understand it correctly that using `LoggingContext` and `logger.context` is essentially the same thing but `LoggingContext` allows the scoping? Again, do we need the `LoggingContext` and do we need to pass it as extra parameter to the class? It seems to me that this will be very cumbersome because you'll have like 2 or 3 constructor parameters yourself and then you'll have two for logging, one for monitoring and who knows what... I'd try to simplify this.

They look similar and even work similarly, but they are two different things. `LoggingContext` is used for context that is not really tied to a single log message. Typical example would be correlation id. On the other hand, `Logger.context` is used for log message specific context. That would typically be some case class.

It might seem that `Logger.context` is not really needed (which in fact is not), but I included it to simplify it for the user. And most importantly, it will allow us to provide more efficient implementation of this use case.

>     * Is there specific reason for the syntax `log.info.msg`? It's a bit unusual to me, I'm more used to `log.info(msg)` or something like that.

Not really, except that I found it more consistent with the rest.

Commit:
cleanup

jveselka commented on core/src/main/scala/slog/LogBuilder.scala:
```diff
@@ -0,0 +1,7 @@
+
+trait LogBuilder[F[_]] {
+  def msg(msg: String): F[Unit]
+  def msg(ex: Throwable)(msg: String): F[Unit]
```
Could it possibly go as far as having separate `msg` and `ex` methods, both aliases to `context` with fixed keys?
```
def msg(msg: => String): LogBuilder[F] = context("message", msg)
def ex(ex: => Throwable): LogBuilder[F] = context("exception", Throwable)
```
Logging itself would not be executed by `msg` method, but with loglevel-determining method e.g. `debug()` or `info()` like @sideeffffect proposed earlier.

hanny24 replied:
I actually had something like this. The problem with it is that it adds at least one more allocation (temporary) object which might be costly.

sideeffffect replied:
I started to like the fluent/builder interface, but maybe we *should* consider using plain methods ala `logger.info("some string", o, p, ex)`

hanny24 replied:
Those methods are very difficult to read. At least for me.

sideeffffect replied:
but less allocations.. I don't really know the cost and don't care that much myself :man_shrugging:

jveselka replied:
Do we really care that much about extra allocations? I understand the concern especially for logging library, but to let it guide API decisions is quite extreme.

Also if we decide to use GraalVM and I understood correctly the magic behind it, these allocations should be eliminated (from heap).

tomasherman replied:
graal won't help you here, most people who use this (even us) won't run this on graal jdk

jakubjanecek commented:
Alright, thanks for answers. So I propose the following:
* Get rid of `LoggerFactory`.
* Change the logging syntax to `log.info(msg)`, I think @sideeffffect is proposing the same thing in one of the comments.
* I'd think about doing only single way of "context" and I would actually do it the other way around so that there is no need for another class/concept. So something like `log.context("key" -> value).info(msg)` for single message and
```scala
log.contextScope("key1" -> value1, "key2" -> value2) {
  ...
}
```

And I realize you'll need a different syntax because you want different value types in the context so maybe something like this:

```scala
log.contextScope("key1" -> value1) {
  log.contextScope("key2" -> value2) {
    ....
  }
}
```
or there could be few `contextScope` methods for one, two, three and maybe four parameters (I doubt anyone will regularly need more than that and if yes there is still the nesting syntax possible).

For me this proposal means that the API will be simpler to use and understand. There is only single concept of `Logger`, the `Logger` can be named (either explicitly or implicitly via `classTag`), it provides context, either locally for single message or as a scope and the usual logging methods, nothing more.

hanny24 commented:
>     * Get rid of `LoggerFactory`.
And replace it with what? 
>     * Change the logging syntax to `log.info(msg)`, I think @sideeffffect is proposing the same thing in one of the comments.
If everyone agrees, I'm OK with it as well.
>     * I'd think about doing only single way of "context" and I would actually do it the other way around so that there is no need for another class/concept. So something like `log.context("key" -> value).info(msg)` for single message and
I disagree with that. There  main reasons for that are:
- It forces you to create a logger even in cases when you don't really need to log something.
- The API somehow suggest that the context is valid only for that particular logger, not for all of them.
- You can always merge two traits into one quite easily, the other way is usually not so simple.
- your proposal might be difficult to implement (both i terms of complexity and performance).

jakubjanecek commented:
I know everyone is doing `LoggerFactory` but that does not mean it's right :) Initially I was proposing to add a method called `named` to `Logger` which would allow you to name it and also adding `implicit classTag` to all the logging methods so that you don't have to worry about the logger name at all. Now I'm thinking if it is possible to use macros to know the enclosing type name... so it would really be easy for the user and you would use the same way of getting line number, filename and logger (type) name.

About the context. You are saying it forces you to create logger even if you don't need to. Alright, I get that but you will have to allocate something anyway (being it logger or loggingcontext) and the way it is used you don't even see what is allocated because it is chained. I'd rather do it that way than have two extra parameters just because of logging in all my classes. I wouldn't say the API suggests that, I don't see it that way, really. The third argument is correct but we are designing an API and you know it will stick that way so I don't that important now. Regarding performance and complexity, I would have to try it myself to actually see but I don't see why my proposal should be much different from the original regarding implementation complexity or performance - meaning you are stating that but you need to explain it if you know for sure it is that way.

Commit:
Merge branch 'master' into init

Commit:
scalafmt

Commit:
add scalafmt cofnig

hanny24 commented:
> I know everyone is doing `LoggerFactory` but that does not mean it's right :) Initially I was proposing to add a method called `named` to `Logger` which would allow you to name it and also adding `implicit classTag` to all the logging methods so that you don't have to worry about the logger name at all. Now I'm thinking if it is possible to use macros to know the enclosing type name... so it would really be easy for the user and you would use the same way of getting line number, filename and logger (type) name.

However there must still be a way how to explicitly provider logger name. `ACCESS_LOG` comes to my mind. What I'm afraid is that with macro it might not be very clear what the result logger name would be. For instance, if we have class nested in some other class.


> About the context. You are saying it forces you to create logger even if you don't need to. Alright, I get that but you will have to allocate something anyway (being it logger or loggingcontext) and the way it is used you don't even see what is allocated because it is chained. I'd rather do it that way than have two extra parameters just because of logging in all my classes. 


I wouldn't say the API suggests that, I don't see it that way, really. The third argument is correct but we are designing an API and you know it will stick that way so I don't that important now. Regarding performance and complexity, I would have to try it myself to actually see but I don't see why my proposal should be much different from the original regarding implementation complexity or performance - meaning you are stating that but you need to explain it if you know for sure it is that way.

I tried to design the API in a way that it can be effectively implemented. For instance, we want to avoid updating (and reverting back) fiber context (`FiberRef`, `TaskLocal`) unless we really need to. Does it make sense?

tomasherman commented on core/src/main/scala/slog/LogBuilder.scala:
```diff
@@ -0,0 +1,7 @@
+package slog
+
+trait LogBuilder[F[_]] {
```
does it ever make sense to have a method that just dumps context without message? for example access logging ... maybe `def nomsg: F[Unit]`

hanny24 replied:
underlying logging might not like it.

sideeffffect commented on core/src/main/scala/slog/LoggingContext.scala:
```diff
@@ -0,0 +1,15 @@
+  def apply[F[_]](implicit ev: LoggingContext[F]): LoggingContext[F] = ev
+
+  trait LoggingBuilder[F[_]] {
+    def scoped[T](fv: F[T]): F[T]
```
another option is `span` (as in tracing labraries). Or `use`, which reminds me of `Resource`, which also has this "enclosed block" meaning.

tomasherman commented on core/src/main/scala/slog/LoggingContext.scala:
```diff
@@ -0,0 +1,15 @@
+
+  trait LoggingBuilder[F[_]] {
+    def scoped[T](fv: F[T]): F[T]
+    def use[T: StructureEncoder](key: String, value: T): LoggingBuilder[F]
```
so we are giving up on context propagation being orthogonal to structured logging?

hanny24 replied:
I don't see a way how can structured logging be implemented without some sort of context propagation. But they can live next to each other just fine.

sideeffffect replied:
@tomasherman I think this library should be focused on logging only, it shouldn't concert with context propagation in general IMHO. at least not in the beginning

:thinking: 
thinking out loud: If we wanted to have general context propagation mechanisms, we probably should have started with that, and build logging (and what else? tracing? correlationID propagation?) on to of it.

tomasherman commented on core/src/main/scala/slog/StructureBuilder.scala:
```diff
@@ -0,0 +1,13 @@
+  def long(value: Long): T
+  def double(value: Double): T
+  def string(value: String): T
+  def structure(name: String, attributes: Map[String, T]): T
```
i would call this map because it's what it is

jakubjanecek commented:
Yes, that's what the `Logger.named` method would be for. But in 99% of cases it would work just fine, I don't think people care much about logger names.

And I believe you are right just I would have to see it in practice to understand what you actually mean. What I don't like about this approach the most is the necessity for another class dependency. Imagine you have a client to some web service. It needs to receive configuration case class and HTTP `Client`, those two things are natural. But on top of that we will now pass logging factory, logging context "factory" and possibly some meter registry and maybe even circuit breaker. I think the business logic gets lost in there. The client is supposed to communicate so it needs `Client` and URL (configuration). It's fine that it requires `Logger` because it probably wants to log something. But why should it require `LoggerFactory` to create just any logger? It does not make sense. That's why I'm trying to push a simpler approach (in my view) which is that the `Logger` is the only thing you need for everything.

Essentially I imagine this (keeping in mind that this is just the basic stuff):
```scala
trait Logger[F[_]] {
  def info(...): F[Unit] // logs a message possibly with Exception at INFO log level, line number, filename and class inferred via macros automatically
  // debug, warn, error...
  def context(...): Logger[F] // "local" context just for single log message
  def scope[A](...)(body: => A): A // creates a scope, all logging will contain data from scope's context and all scopes higher in the stack
  def named(loggerName: String): Logger[F] // names the logger and the automatic naming via macros is "disabled"
}
```
Why is this bad?

tomasherman commented on core/src/main/scala/slog/StructureEncoder.scala:
```diff
@@ -0,0 +1,46 @@
+package slog
+
+trait StructureEncoder[T] {
+  def encode[O](value: T)(implicit structureBuilder: StructureBuilder[O]): O
```
i would use context obunds here , no need to give it a name ;)

hanny24 replied:
not important right now

tomasherman commented on example/src/main/scala/example/Example.scala:
```diff
@@ -0,0 +1,66 @@
+    logger.info.msg("Hello world!!") >>
+      // Provide extract context to the log message
+      // {"message": "Hello", "file": "Example.scala", "line": 22, "correlation_id": "<VALUE>"}
+      logger.info.context("correlation_id", "<VALUE>").msg("Hello") >>
```
i wonder if this would be a good usecase for `HList` ... calling `context` all the time adds a lot of visual noise - having `HList[(String, SomethingSerializable)]` might be the way to go

hanny24 replied:
You still need to provide keys. And shapeless is not super user friendly.

sideeffffect replied:
I also think we should have fluent interface rather than HList because of use-friendliness

tomasherman commented on example/src/main/scala/example/Example.scala:
```diff
@@ -0,0 +1,66 @@
+        .context("correlation_id", "<VALUE>")
+        .msg(new Exception)("Hello") >>
+      loggingContext
+        .use("correlation_id", "<VALUE>")
```
same as the above for `HList`

tomasherman commented:
one thing that i dislike on the `named` approach is that it forces you to hunt through the sourcecode for the actually logger name which you will when you're trying to configure `logback.xml` levels - with regular approach with 1:1 mapping to classname it's obvious, with nammed - not so much ... and you are basically forcing people to come up with their own convention - every time in every service

jakubjanecek commented:
This is how I think it could work: https://gist.github.com/jakubjanecek/25c929b5e8d6d4e2e707825d61ecc6c4

sideeffffect commented:
To sum up my thoughts
1) Custom encoders: I tend to be against, but I seem to be in minority. people either don't care or want custom
2) Logging and context augmentation: Only major outstanding issue now. I prefer having this in one object, mostly for simplicity

@jakubjanecek I think we're actually converging, if you have a look at my take on Structured logging for Scala https://github.com/sideeffffect/loggingexperiment/pull/2/files you'll see we're not that far from each other :wink: 
Also anybody else feel free to check it out as inspiration/food for thought before Monday meeting, it comes with a working implementation too

hanny24 commented:
I'm currently have one more idea, that is little bit different. How about this:
```scala
logger.info(log"Test".exception(new RuntimeException).context("correlation_id","<VALUE>"))
```
 From my POV it has two benefits: first, it's a little bit cleaner that the context is related to the message itself. Secondly and most importantly it allows us to provide really efficient implementation (in a sense that we active log level check is as efficient as hand written code)

Commit:
progress

Commit:
when enabled

Commit:
when enabled improvements

Commit:
small improvements

hanny24 commented:
So I pushed few changes. To sum it up:
- Renamed few methods such that it aligns with offline discussion.
- introduced `logger.level.whenEnabled` that can be used to compute logging arguments only when log level is enabled. There is one single method that needs to be implemented for this functionality.  There are a bunch of helper functions provided on top of it.
- There are `LogBuilder`, `LevelLogBuilder` and `WhenEnabledLogBuilder` traits. They might seem scary at first, but regular user does not really need to know the difference. Disclaimer: the reason for them is to provide API that allows user to do only "allowed" operations (ie. it doesn't make sense to check log level twice)
- The whole API is designed such that logger implementations are relatively easy. Also, we limit macro usage as much as possible. No more macros should be needed apart the one that are already implemented.

jakubjanecek commented:
LGTM

Few things:
* I don't see a reason for currying when exception is logged so simple `logger.error.log(ex, "msg")` would seem better to me.
* I would add the `log` method even for the simple logging case and make `apply` just an alias - seems more clear to me that way.
* I would move the different `traits` to separate files etc. - I guess this is just work in progress.

Otherwise I think we've come to a good API 🎉

sideeffffect commented on src/main/scala/slog/Logger.scala:
```diff
@@ -0,0 +1,13 @@
+
+import slog.macros.LoggerImpl
+
+class Logger[F[_]](val underlying: LocationAwareLogger[F]) {
```
I suppose the `LocationAwareLogger` is a stand in for `slf4j` `Logger`, is that correct?

hanny24 replied:
Not really. It's actual logger implementation that doesn't need any macros.

sideeffffect commented on example/src/main/scala/example/Example.scala:
```diff
@@ -0,0 +1,84 @@
+class Example[F[_]](
+    loggerFactory: LoggerFactory[F],
+    loggingContext: LoggingContext[F]
+)(implicit F: Monad[F]) {
```
can we please have a type which is a subtype of `Logger` and `LoggingContext`, that will be used here in the example?

hanny24 replied:
I thought we agreed not to include it in the first version.

jakubjanecek replied:
I think so too. That can always be added once we have the base APIs ready and agreed upon :)

sideeffffect commented on core/src/main/scala/slog/macros/LoggerImpl.scala:
```diff
@@ -0,0 +1,37 @@
+    import c.universe._
+    val file = Literal(Constant(c.enclosingPosition.source.path.toString))
+    val line = Literal(Constant(c.enclosingPosition.line))
+    val tree = q"${c.prefix}.underlying.$level($file, $line)"
```
I strongly against this. This is the exact opposite of what I had in mind when I stated the requirement that the file name and line number should be present in the log (and maybe also other things, line method and class names).

We should delegate that to the underlying `slf4j` (or `Logback`)

hanny24 replied:
I'm afraid that it is not feasible in a maintainable way. That would require every logger implementation to be macro based, which is in my perspective not a good idea.

jakubjanecek replied:
I've said it before. Even though I understand Ondra's point I do not agree with it in this particular case. Moreover, it does not seem much important, it's just two lines of code which can be removed if we find out during implementation and first alpha releases that it shouldn't be there or the other way around. For now I would leave it and let's see.

hanny24 replied:
This is not just "two lines of code". It's more fundamental issue. What @sideeffffect proposes is essentially getting rid of traits and use single class implementation only. I think that this is not really a way forward as it would would make the library not so useful and hardly extensible.

sideeffffect commented:
> It's more fundamental issue

I'm worried this proposal is getting way to complicated. please, let me think about it :pray:

Commit:
slf4j implementation

Commit:
logging context

Commit:
example

Commit:
tests

Commit:
noop

Commit:
scalafmt

Commit:
Better generic

Commit:
github

hanny24 commented:
I decided to merge this for now such that we can move forward. We can still change things later on.

sideeffffect commented:
Unfortunately as this is now, it's of no use to us. It has developed into a new full-fledged framework that does too many things tailored to your particular use-case.
I would like to propose https://github.com/sideeffffect/loggingexperiment as a different proposal -- it's smaller and more focused, and I think it solves some of the issues in a bit more elegant way. (But of course, took a lot of inspiration from this `slog` proposal.)
